packages/services/src/kernel/default.ts

```diff
@@ -31,6 +31,10 @@ import * as restapi from './restapi';
 // Stub for requirejs.
 declare let requirejs: any;
 
+const RESTARTING_KERNEL_SESSION = '_RESTARTING_';
+
+let _pendingMessages: KernelMessage.IMessage[] = [];
```

Comment:
As I read it, any queued messages would go to the first kernel that starts.   So, if you tried to run cells from two different notebooks while their kernels were starting then you'd end up executing both sets of code on one kernel.

Reply:
Nothing escapes your eagle eyes @blink1073 :)  I have created that static field (while debugging...) but reverted it back to an instance field in this https://github.com/jupyterlab/jupyterlab/pull/8562/commits/f0a1a1a767f54b6f767ef0d91c89aad36e31683c .... but forgot to push it. Now the `_pendingMessages` is again an instance of the `KernelConnection`.

Anyway, I have just tested so far with one restarting kernel, so yeah, I will test with multiple restarting kernels to ensure we meet the expected behavior.

Reply:
I have done a few tests and the pendingMessages do not mangle each other in case of multiple slow kernels restarting at the same time. I have also opened https://github.com/jupyter/jupyter_server/issues/247 where we can implement a better solution for this feature. For now, I believe the proposed solution is "good enough".